### PR TITLE
common,config: OPT_FLOAT and OPT_DOUBLE output format in config show

### PR DESCRIPTION
**Pull request by gmayyyha**
Fixes: http://tracker.ceph.com/issues/20104

Signed-off-by: Yanhu Cao <gmayyyha@gmail.com>

### CODE REVIEW

File: src/common/config.cc [hunk 1]
```diff
@@ -949,6 +953,10 @@ int md_config_t::_get_val(const char *key, char **buf, int len) const
     ostringstream oss;
     if (bool *flagp = boost::get<bool>(&cval)) {
       oss << (*flagp ? "true" : "false");
+    } else if (float *fp = boost::get<float>(&cval)) {
```
Comment (tchaikov) [inline src/common/config.cc]: can we refactor this method to reuse `md_config_t::_get_val(const char *key, std::string *value)`?

Comment (gmayyyha) [inline src/common/config.cc]: like this ?
```CPP
@@ -946,33 +946,21 @@ int md_config_t::_get_val(const char *key, char **buf, int len) const
   if (!key)
     return -EINVAL;
 
-  string k(ConfFile::normalize_key_name(key));
-
-  config_value_t cval = _get_val(k.c_str());
-  if (!boost::get<invalid_config_value_t>(&cval)) {
-    ostringstream oss;
-    if (bool *flagp = boost::get<bool>(&cval)) {
-      oss << (*flagp ? "true" : "false");
-    } else if (float *fp = boost::get<float>(&cval)) {
-      oss << std::fixed << *fp ;
-    } else if (double *dp = boost::get<double>(&cval)) {
-      oss << std::fixed << *dp ;
-    } else {
-      oss << cval;
-    }
-    string str(oss.str());
-    int l = strlen(str.c_str()) + 1;
+  string val ;
+  if (!_get_val(key, &val)) {
+    int l = strlen(val.c_str()) + 1;
     if (len == -1) {
       *buf = (char*)malloc(l);
       if (!*buf)
         return -ENOMEM;
-      strcpy(*buf, str.c_str());
+      strcpy(*buf, val.c_str());
       return 0;
     }
-    snprintf(*buf, len, "%s", str.c_str());
+    snprintf(*buf, len, "%s", val.c_str());
     return (l > len) ? -ENAMETOOLONG : 0;
-  }
+  } 
 
+  string k(ConfFile::normalize_key_name(key));
```

Comment (tchaikov) [inline src/common/config.cc]: something like that, we can have two commits here. one for the refactor, the other for fixing the output format.

also, you don't need to use `strlen()` for the length of a `std::string`. you could just use `val.length()` instead. and please use `strncpy()` if possible.

File: src/common/config.cc [hunk 2]
```diff
@@ -946,33 +946,21 @@ int md_config_t::_get_val(const char *key, char **buf, int len) const
...
-    } else if (double *dp = boost::get<double>(&cval)) {
-      oss << std::fixed << *dp ;
-    } else {
-      oss << cval;
-    }
-    string str(oss.str());
-    int l = strlen(str.c_str()) + 1;
+  string val ;
+  if (!_get_val(key, &val)) {
+    int l = val.length()+1;
```
Comment (tchaikov) [inline src/common/config.cc]: please add spaces around `+`

Comment (tchaikov) [inline src/common/config.cc]: and could you reorder these two commits? so the refactor one is the first commit, and the fix is the second.

Comment (gmayyyha) [inline src/common/config.cc]: spaces are already used

qq20170614-141954

Comment (tchaikov) [inline src/common/config.cc]: could you check again?

```c++
int l = val.length()+1;
```

File: src/common/config.cc [hunk 3]
```diff
@@ -942,29 +942,21 @@ int md_config_t::_get_val(const char *key, char **buf, int len) const
...
         return -ENOMEM;
-      strcpy(*buf, str.c_str());
+      strncpy(*buf, val.c_str(), l);
       return 0;
     }
-    snprintf(*buf, len, "%s", str.c_str());
+    snprintf(*buf, len, "%s", val.c_str());
     return (l > len) ? -ENAMETOOLONG : 0;
-  }
+  }
```
Comment (tchaikov) [inline src/common/config.cc]: and please remove the trailing space.

File: src/common/config.cc (hunk 2 above)
Comment (gmayyyha) [inline src/common/config.cc]: is this correct now ?

Comment (tchaikov) [inline src/common/config.cc]: no. @gmayyyha you can review your commit with the github web UI by yourself as well, see https://github.com/ceph/ceph/pull/15647/files

File: src/common/config.cc (hunk 3 above)
Comment (gmayyyha) [inline src/common/config.cc]: @tchaikov and now ?

File: src/common/config.cc [hunk 4]
```diff
@@ -942,29 +946,21 @@ int md_config_t::_get_val(const char *key, char **buf, int len) const
...
-      strcpy(*buf, str.c_str());
+      strncpy(*buf, val.c_str(), l);
       return 0;
     }
-    snprintf(*buf, len, "%s", str.c_str());
+    snprintf(*buf, len, "%s", val.c_str());
     return (l > len) ? -ENAMETOOLONG : 0;
   }
 
+  string k(ConfFile::normalize_key_name(key));
```
Comment (Liuchang0812) [inline src/common/config.cc]: `k` is useless. remove this line.

File: src/common/config.cc [hunk 5]
```diff
@@ -942,29 +946,21 @@ int md_config_t::_get_val(const char *key, char **buf, int len) const
...
       *buf = (char*)malloc(l);
       if (!*buf)
         return -ENOMEM;
-      strcpy(*buf, str.c_str());
+      strncpy(*buf, val.c_str(), l);
       return 0;
     }
-    snprintf(*buf, len, "%s", str.c_str());
+    snprintf(*buf, len, "%s", val.c_str());
     return (l > len) ? -ENAMETOOLONG : 0;
```
Comment (Liuchang0812) [inline src/common/config.cc]: @tchaikov hi, kefu, I'm confused here. We should allocate memory when `len` is `-1`, and we will always return ENAMETOOLONG error code. is this what we expect?

File: src/common/config.cc [hunk 6]
```diff
@@ -942,29 +946,21 @@ int md_config_t::_get_val(const char *key, char **buf, int len) const
...
-    int l = strlen(str.c_str()) + 1;
+  string val ;
+  if (!_get_val(key, &val)) {
+    int l = val.length() + 1;
     if (len == -1) {
       *buf = (char*)malloc(l);
       if (!*buf)
         return -ENOMEM;
-      strcpy(*buf, str.c_str());
+      strncpy(*buf, val.c_str(), l);
```
Comment (Liuchang0812) [inline src/common/config.cc]: we could simplify line 956 and 959 to one line actually.  anyway, I'm fine with both.

File: src/common/config.cc (hunk 4 above)
Comment (gmayyyha) [inline src/common/config.cc]: @Liuchang0812 `k` is used as follows
```c++
965   for (int o = 0; o < subsys.get_num(); o++) {
966     std::string as_option = "debug_" + subsys.get_name(o);
967     if (k == as_option) {
```

Comment (Liuchang0812) [inline src/common/config.cc]: ok. good.

File: src/common/config.cc (hunk 5 above)
Comment (tchaikov) [inline src/common/config.cc]: > We should allocate memory when `len` is `-1`, 

correct.

> and we will always return ENAMETOOLONG error code. is this what we expect?

i don't think so.

Comment (Liuchang0812) [inline src/common/config.cc]: I created a issue to track this issue[1]. I will write some UT to verify it. 

1: http://tracker.ceph.com/issues/20442

Comment (Liuchang0812) [inline src/common/config.cc]: I missed line957.  _get_val will not return ENAMETOOLONG when len is -1.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-06_22:05:50-rados-wip-sage-testing2-distro-basic-smithi/

**Pull request merged**